js/imageeditor/shared/tools/brush.ts

```diff
@@ -0,0 +1,246 @@
+	/**
+	 * The opacity of the brush.
+	 */
+	opacity: number;
+	/**
+	 * Whether or not to set the initial texture.
+	 */
+	set_initial_texture?: boolean;
+}
+
+/**
+ * Draws a circle on the given graphics object.
+ * @param graphics the graphics object to draw on
+ * @param x the x coordinate of the circle
+ * @param y the y coordinate of the circle
+ * @param brush_color the color of the circle
+ * @param brush_size the radius of the circle
+ */
+function drawCircle(
+	graphics: Graphics,
+	x: number,
+	y: number,
+	brush_color: ColorSource = new Color("black"),
+	brush_size: number
+): void {
+	const color = new Color(brush_color);
+	graphics.beginFill(color);
+	graphics.drawCircle(x, y, brush_size);
+	graphics.endFill();
+}
+
+/**
+ * Interpolates between two points.
+ * @param point1 the first point
+ * @param point2 the second point
+ * @returns an array of points between the two points
+ */
+function interpolate(
+	point1: { x: number; y: number },
+	point2: { x: number; y: number }
+): Points[] {
+	let points: Points[] = [];
+	const dx = point2.x - point1.x;
+	const dy = point2.y - point1.y;
+	const distance = Math.sqrt(dx * dx + dy * dy);
+	const steps = Math.ceil(distance / 2);
+	const stepX = dx / steps;
+	const stepY = dy / steps;
+
+	for (let j = 0; j < steps; j++) {
+		const x = point1.x + j * stepX;
+		const y = point1.y + j * stepY;
+		points.push({ x, y });
+	}
+
+	return points;
+}
+
+export function draw_path(
+	renderer: IRenderer,
+	stage: Container,
+	layer: LayerScene,
+	mode: "draw" | "erase"
+): DrawCommand {
+	const paths: Points[] = [];
+	let initial_path: DrawOptions;
+	let graphics: Graphics;
+	let InitialTexture: RenderTexture;
+
+	let has_drawn = false;
+	let id = 0;
+
+	return {
+		drawing: false,
+		start: function ({
+			x,
+			y,
+			size,
+			color = new Color("black"),
+			opacity,
+			set_initial_texture = true
+		}: DrawOptions) {
+			if (set_initial_texture) {
+				InitialTexture = RenderTexture.create({
+					width: layer.draw_texture.width,
+					height: layer.draw_texture.height
+				});
+				renderer.render(layer.composite, {
+					renderTexture: InitialTexture
+				});
+			}
+			initial_path = { x, y, size, color, opacity };
+			paths.push({ x, y });
+			graphics = make_graphics(id++);
+
+			drawCircle(graphics, x, y, color, size);
+
+			renderer.render(graphics, {
+				renderTexture:
+					mode === "draw" ? layer.draw_texture : layer.erase_texture,
+				clear: false
+			});
+
+			this.drawing = true;
+		},
+		continue: function ({ x, y }: Points) {
+			const last_point = paths[paths.length - 1];
+			const new_points = interpolate(last_point, { x, y });
+
+			for (let i = 0; i < new_points.length; i++) {
+				const { x, y } = new_points[i];
+				drawCircle(graphics, x, y, initial_path.color, initial_path.size);
+				paths.push({ x, y });
+			}
+
+			renderer.render(graphics, {
+				renderTexture:
+					mode === "draw" ? layer.draw_texture : layer.erase_texture,
+				clear: false
+			});
```

Comment:
Not sure if it is the interpolation that is causing issues or just rendering large textures.